### DISCUSSION TITLE
Categorical metadata for timeseries data

### DISCUSSION THREAD

**Discussion by geoHeil** [timeseriesAI/tsai#163] [Q&A]
When I have categorical metadata (let`s assume federal states or similar attributes) how can I feed this additionally to Tsai?

One I.e. how can I ideally have Tsai calculate an embedding for the categories and concatenate the embeddings with the slide time-series data and then finally feed the whole data matrix to the neural network.

**Comment by vrodriguezf** [comment]
If you have tabular metadta, you can try concating those features at the end of the neural network with a `MixedDataLoader` and a `MultiInputNet`:
https://timeseriesai.github.io/tsai/models.MultiInputNet.html#MultiInputNet

You can try to add the [Gradient Blending callback](https://github.com/timeseriesAI/tsai/blob/main/nbs/062_callback.gblend.ipynb) to that learner and see if that improves learning better from each modality.

**Reply by geoHeil** [reply]
`get_mixed_dls` - does not seem to produce a JOIN between the meta information of each category and the time series. How is this joining up happening implicitly?

**Reply by geoHeil** [reply]
https://github.com/timeseriesAI/tsai/blob/main/nbs/062_callback.gblend.ipynb in `ts_features_df = get_ts_features(X, y)` is using tsfresh to calculate the non temporal features. Howeve, it is creating an ID (if I am not mistaken for each series individually and not for a group). So the shape of the output features should be in the dimension of #series. But if I have regions like federal states with multiple devices within each state (which are producing time series) where/how is the join of this information happening?

**Reply by oguiza** [reply]
Hi @geoHeil,
I'm not sure what you mean by the "meta information of each category and the time series". Could you please explain what you mean by that?

**Reply by geoHeil** [reply]
i.e. if one time series belongs to federal state A and I know that A has certain lat/long coordinates or sociodemographics. I.e. static information which does not change in the observation period but is specific to one of the categories.

**Reply by oguiza** [reply]
Thanks. I think I understand your question now. 

In a mixed dataloader, the information is never joined. You  need to prepare the data so you can feed them to 2 different dataloaders: 
1. A numpy array (or array-like) with the time series (and any other dynamic covariates if available) into a time series dataloader (you can use get_ts_dls)
2. A pandas df with the static covariates (like categorical or continuous data) into a tabular dataloader (you can use get_tabular_dls)

Both sets of data must have the same order when you pass them to the dataloaders. 

What the mixed dataloader will do is create a batch that will contain a tuple with the time series, categorical and continuous data. In the train set, the indices applied to both dataloaders will be the same. So even if you shuffle the data, they will always belong to the same entity. In addition to that, the mixed dataloader will create a single target (that is the same for both dataloaders). The tuple will be passed to the model. You can use a MultiInputNet to take those tuples. 

There is an example in the [documentation](https://timeseriesai.github.io/tsai/models.MultiInputNet.html).
I hope this is clearer now.

**Reply by geoHeil** [reply]
You mention the data is never joined - but this is exactly my question: Does this mean you assume for each timeseries (index) from the TS dls an equal index in the static covariates DLS? But then I anyways need to join it first only to separate it later.

**Reply by oguiza** [reply]
Sorry for the late reply. 

Yes, you are correct. The assumption is that all data that you pass to multiple dataloaders are indexed in the same way.

I understand that this may not be the ideal solution, but it provides a lot of flexibility. It's possible to build a dataloader for a particular case, but I can't do that with a library that is used with many different types of inputs.
